### PR TITLE
Phase 3 Task 13 cycle 2: <thead> / <tfoot> repetition at page boundaries

### PR DESCRIPTION
**Pull request by raroche**
## Summary

Task 13 cycle 1 ([PR #54](https://github.com/raroche/NetPdf/pull/54)) shipped row-level pagination via `TableContinuation`; the `RepeatHead` / `RepeatFoot` fields on the continuation were reserved for cycle 2. Cycle 2 ships **CSS Tables L3 §3.6 / §11 header + footer repetition**: when a `<thead>` (TableHeaderGroup) or `<tfoot>` (TableFooterGroup) exists, those rows now **repeat at the top + bottom of EACH page** where the table appears (not just page 1).

Continues from PR #54 merged at `8c84c30`.

## Row classification

`CollectRows` extended to tag each row with its source group kind (`Header` / `Body` / `Footer`). BoxBuilder's table fixup already reorders `<tfoot>` to the end per HTML5; cycle 2 relies on that pre-sort. Row slices:
- `[0, headerCount)` — header rows
- `[headerCount, rows.Count - footerCount)` — body rows
- `[rows.Count - footerCount, rows.Count)` — footer rows

## New fields

`_measuredHeaderStackHeight` + `_measuredFooterStackHeight` cache the cumulative row heights for the header/footer slices — stable across pages since the same headers/footers repeat.

## Per-page emit phases (5 → 7)

| Phase | Content | When |
|---|---|---|
| **0** | Top captions | First page only |
| **A-pre** | Header rows | Every page (if `_headerRowCount > 0`) |
| **A** | Body rows (cycle 1's pagination loop) | Bounded to `[headerCount, rows.Count - footerCount)` |
| **B / C** | Cells + content | Interleaved with body row emit |
| **A-post** | Footer rows | Every page (if `_footerRowCount > 0`), immediately after last body row |
| **D** | Bottom captions | Last page only |

## Body block budget reservation

Before the body-row pagination loop: `fragmentainer.UsedBlockSize = rowStackOrigin + headerStackHeight + footerStackHeight`. Body row chunk-size unchanged; the resolver naturally sees the reduced remaining space (= `BlockSize - headerStackHeight - footerStackHeight`). `UsedBlockSize` restored to `initialUsedBlockSize` after the loop so the outer BlockLayouter's cursor advance doesn't double-count.

## Footer positioning

Footers emit **immediately after** the last body row that fit on this page (NOT anchored at absolute page bottom). Sub-cycle 3+ may add bottom-anchoring if needed.

## `RepeatHead` / `RepeatFoot` flags

- **First page** (`_incomingTableContinuation == null`): `repeatHead` / `repeatFoot` = `false` (headers emit naturally).
- **On `PageComplete`**: new `TableContinuation` built with `RepeatHead = (_headerRowCount > 0)` + `RepeatFoot = (_footerRowCount > 0)` — signals "I have headers/footers that need to repeat".
- **On resume page** with `RepeatHead == true`: Phase A-pre fresh-emits headers at row-stack-origin. Same for `RepeatFoot`.

## New diagnostic

`LAYOUT-TABLE-HEADER-FOOTER-OVERSIZED-001` Warning — catastrophic case where `headerStackHeight + footerStackHeight > fragmentainer.BlockSize - rowStackOrigin` (no room for any body content). Cycle 2 still emits headers + footers but skips body rows on that page (avoids infinite-resume).

Added to `PaginateDiagnosticCodes` + facade + `docs/diagnostics-codes.md` + 3 parity tests.

## Edge cases handled

- **No `<thead>` / `<tfoot>`**: cycle 1 behavior preserved.
- **Header taller than fragmentainer**: atomic + existing `PAGINATION-FORCED-OVERFLOW-001`.
- **Footer taller than fragmentainer**: similar.
- **Header + footer combined > fragmentainer**: new `LAYOUT-TABLE-HEADER-FOOTER-OVERSIZED-001`.
- **Table with only `<thead>` / only `<tfoot>`**: emit + `AllDone`.
- **Top captions on resume pages**: skipped (cycle 1 preserved).
- **Bottom captions** follow the footer on the last page.

## Caveat

Under `<html><body><table>` the **recursive-atomic fallback** from cycle 1 Finding 1 keeps the table on a single page; cycle 2's per-page repetition is fully exercised by direct-construction unit tests. Sub-cycle 6+ may generalize `EmitBlockSubtreeRecursive` to propagate continuations end-to-end.

## Files changed

8 source/test files + 2 docs:

- **`src/NetPdf.Layout/Layouters/TableLayouter.cs`** — +866/−119: row classification, header/footer height fields, Phase A-pre / A-post emit, body budget reservation, `RepeatHead`/`RepeatFoot` consumption.
- **`src/NetPdf.Paginate/Diagnostics/PaginateDiagnosticCodes.cs`** + **`src/NetPdf/Diagnostics/DiagnosticCodes.cs`** — new `LayoutTableHeaderFooterOversized001` constant.
- **`tests/NetPdf.UnitTests/Phase3/TableLayouterTests.cs`** — +533: 11 new direct cycle-2 tests.
- **`tests/NetPdf.UnitTests/Phase3/TableLayouterProductionTests.cs`** — +59: 1 new production test.
- **`tests/NetPdf.UnitTests/Diagnostics/DiagnosticCodesTests.cs`** — +31: 3 parity tests for new diagnostic.
- **`docs/deferrals.md`** — Behavior describes header/footer repeat; Missing drops "per-page header/footer repeat (Task 13)".
- **`docs/diagnostics-codes.md`** — new diagnostic row.
- **`PROGRESS.md`** — cycle 2 entry.

Net diff: **+1456 / −121** across 9 modified files.

## Tests

**4304 unit / 4435 solution-wide passing** (+15 unit / +15 solution-wide over PR #54's 4289 / 4420 baseline; 4 skipped). AOT/JIT byte-parity gate stays green; `git diff --check` clean.

New direct tests in `TableLayouterTests.cs`:

1. `Cycle2_table_with_thead_repeats_header_on_each_page`
2. `Cycle2_table_with_tfoot_repeats_footer_on_each_page`
3. `Cycle2_table_with_thead_and_tfoot_repeats_both`
4. `Cycle2_continuation_repeathead_repeatfoot_flags_set_correctly`
5. `Cycle2_no_thead_or_tfoot_preserves_cycle_1_behavior`
6. `Cycle2_table_with_only_thead_emits_only_headers`
7. `Cycle2_table_with_only_tfoot_emits_only_footers`
8. `Cycle2_table_with_oversized_header_emits_forced_overflow`
9. `Cycle2_table_with_header_and_footer_exceeding_fragmentainer_emits_diagnostic`
10. `Cycle2_top_caption_skipped_on_resume_pages`
11. `Cycle2_bottom_caption_emitted_only_on_last_page_after_footer`

New production test: `Cycle2_production_html_table_with_thead_tfoot_emits_header_and_footer_fragments`

## Sub-cycle 3+ / sub-cycle 6+ scope (still deferred)

- Footer bottom-anchoring (cycle 2 emits footer immediately after last body row)
- Recursive-emission continuation propagation past single-level (cycle 1 Finding 1 deferral)
- Row-internal splitting (single oversized row falls back to forced-overflow)
- Spec-strict §11 rowspan distribution-proportional algorithm across pages
- Per-row checkpoint capture for `break-inside: avoid` (cycle 1 Finding 5 deferral)
- Plus all remaining Task 12 sub-cycle 6+ items (border-collapse, percentages, etc.)

## Test plan

- [x] `dotnet build NetPdf.slnx -c Release` — 0 warnings, 0 errors
- [x] `dotnet test NetPdf.slnx -c Release` — 4304 unit / 4435 solution-wide passing
- [x] `./scripts/aot-parity.sh` — AOT/JIT byte-parity verified
- [x] `git diff --check` — clean (no trailing whitespace)
- [ ] Roland review cycle on this PR
- [ ] Address Copilot findings if any

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/NetPdf.Layout/Layouters/TableLayouter.cs [hunk 1]
```diff
@@ -1132,6 +1252,41 @@ public LayoutAttemptResult AttemptLayout(
...
+                    PaginateDiagnosticCodes.LayoutTableHeaderFooterOversized001,
+                    $"TableLayouter: combined <thead> + <tfoot> stack height "
+                    + $"({(headerStackHeight + footerStackHeight):0.##} = "
+                    + $"header {headerStackHeight:0.##} + footer {footerStackHeight:0.##}) "
+                    + $"exceeds the fragmentainer's available block-size "
+                    + $"({(fragmentainer.BlockSize - rowStackOrigin):0.##}). "
+                    + "No room for a body row on a page that also honors the "
+                    + "per-page repeat contract. The layouter commits the "
+                    + "header + footer once on the current page + skips the "
+                    + "body to avoid an infinite continuation loop. Reduce "
```
[the remark this answers is not in the document]
Comment (raroche) [reply src/NetPdf.Layout/Layouters/TableLayouter.cs:1380]: Resolved in (this hardening commit) as Roland's Finding #2 — the contract is now enforced. When `headerStackHeight + footerStackHeight > availableBlockSize` (the catastrophic case), the layouter emits headers + footers atomically and returns `LayoutAttemptResult.AllDone` WITHOUT entering the body loop. Test `Cycle2_table_with_header_and_footer_exceeding_fragmentainer_skips_body_rows` (replaces the prior diagnostic-only assertion) verifies: (a) the diagnostic fires, (b) NO TableRow fragments are emitted for body rows, (c) the result is `AllDone`.

File: src/NetPdf.Layout/Layouters/TableLayouter.cs [hunk 2]
```diff
@@ -1631,6 +1911,192 @@ private void EmitRowWindow(
...
+        for (var r = headerStart; r < headerEndExclusive; r++)
+        {
+            var bucket = placementsByRow[r];
+            if (bucket is null) continue;
+            cancellationToken.ThrowIfCancellationRequested();
+            for (var i = 0; i < bucket.Count; i++)
+            {
+                var placement = bucket[i];
+                placement.ContentBuffer.FlushKeepingBuffer(_sink, rowBlockOffsets[r]);
+            }
```
[the remark this answers is not in the document]
Comment (raroche) [reply src/NetPdf.Layout/Layouters/TableLayouter.cs:2217]: Resolved in (this hardening commit) as Roland's Finding #3. New per-instance `_diagnosticsFlushedOnce` flag prevents duplication: first-page TableLayouter drains the header-cell `DiagnosticsBuffer` once on its first header emit; resume-page layouters have the flag set TRUE at construction so the prior-already-drained buffer doesn't flush again. `FlushKeepingBuffer` continues to preserve the cell content for the repeat (unchanged). New test `Cycle2_header_cell_diagnostics_flush_once_across_multi_page` pins the behavior — header cell triggers a diagnostic on first emit; multi-page table; assert the diagnostic appears exactly once across both pages.

File: src/NetPdf.Layout/Layouters/TableLayouter.cs [hunk 3]
```diff
@@ -1631,6 +1911,192 @@ private void EmitRowWindow(
...
+        for (var r = footerStartIndex; r < footerEndExclusive; r++)
+        {
+            var bucket = placementsByRow[r];
+            if (bucket is null) continue;
+            cancellationToken.ThrowIfCancellationRequested();
+            for (var i = 0; i < bucket.Count; i++)
+            {
+                var placement = bucket[i];
+                placement.ContentBuffer.FlushKeepingBuffer(_sink, rowBlockOffsets[r]);
+            }
```
[the remark this answers is not in the document]
Comment (raroche) [reply src/NetPdf.Layout/Layouters/TableLayouter.cs:2305]: Resolved in (this hardening commit) as Roland's Finding #3 — same fix as the header-cell case (Copilot #2). The `_diagnosticsFlushedOnce` flag covers both header AND footer cells uniformly. The footer-cell test analog asserts the same flush-once semantics.

File: src/NetPdf.Layout/Layouters/TableLayouter.cs [hunk 4]
```diff
@@ -1867,8 +2362,17 @@ public double MeasureContentHeight(
...
+        // each row's source group kind (Header / Body / Footer) so the
+        // emit phase can re-emit headers + footers on every page. The
+        // collector also reorders so headers come first + footers last
+        // regardless of HTML5 source order (HTML5 permits <tfoot> before
+        // <tbody>; the spec says the footer still renders at the end).
         var rows = new List<Box>();
-        CollectRows(grid, rows, cancellationToken);
+        var rowKinds = new List<RowGroupKind>();
+        CollectRows(grid, rows, rowKinds, out var headerCount, out var footerCount, cancellationToken);
+        _measuredHeaderRowCount = headerCount;
```
[the remark this answers is not in the document]
Comment (raroche) [reply src/NetPdf.Layout/Layouters/TableLayouter.cs:2618]: Resolved in (this hardening commit) as Copilot #4 cleanup. The `rowKinds` array at line 2374 was truly dead after `CollectRows` returned — only `headerCount` / `footerCount` counts were consumed downstream. The array allocation has been removed. Verified by running the full test suite (4308 unit + 4439 solution-wide passing) after removal.

File: PROGRESS.md [hunk 5]
```diff
@@ -1,13 +1,15 @@
...
...+ body + footer rows all flow through BoxBuilder + emit as TableRow fragments). AOT/JIT byte-parity gate stays green; `git diff --check` clean. **Docs:** `docs/deferrals.md`'s `table-auto-fixed-spans-borders` entry's Behavior gains a paragraph describing the header/footer repeat algorithm; Missing drops "per-page header/footer repeat (Task 13)"; retains all the remaining sub-cycle 6+ items (border-collapse, percentages, spec-strict §11 rowspan, RTL, recursive continuation propagation, etc.). **Sub-cycle 3+ / sub-cycle 6+ scope (still deferred):** footer bottom-anchoring (cycle 2 emits footer immediately after last body row); recursive-emission continuation propagation past single-level (cycle 1 Finding 1 deferral); row-internal splitting; spec-strict §11 rowspan distribution-proportional algorithm across pages; per-row checkpoint capture for `break-inside: avoid` (cycle 1 Finding 5 deferral); plus all remaining Task 12 sub-cycle 6+ items (border-collapse, percentages, etc.). Active-task pointer advances to **Task 14** — `MulticolLayouter` — OR **Task 12 sub-cycle 6** — border-collapse + percentage widths + spec-strict proportional-weight distribution, depending on review priority.)
```
[the remark this answers is not in the document]
Comment (raroche) [reply PROGRESS.md]: Resolved in (this hardening commit) as Copilot #5. **You're right** — `GridLevelFixup` (`src/NetPdf.Layout/Boxes/BoxBuilder.cs:1019`) preserves source order; it does NOT reorder `<tfoot>` to the end of the row sequence. The reordering happens INSIDE `TableLayouter.CollectRows` itself: after walking grid children in document order into per-kind buffers (`headers`, `bodies`, `footers`), the final row list is assembled in headers → body → footers order. This was an incorrect claim in the PROGRESS.md entry + the implementation comment. PROGRESS.md updated to accurately describe the in-TableLayouter reordering. New test `Cycle2_table_with_tfoot_before_tbody_in_source_renders_footer_at_end` pins the behavior end-to-end (HTML5 allows `<tfoot>` before `<tbody>` in source order; this is critical for correct rendering).
